### DISCUSSION TITLE
Kong DNS Resolution Order Proposal/Enhancement.

### DISCUSSION THREAD

**Discussion by jeremyjpj0916** [Kong/kong#8693] [Ideas and feature requests]
Why is the default:

```
LAST,SRV,A,CNAME
```

And would Kong have a problem with adjusting the default to something like:

```
LAST,A,SRV,CNAME
```

Because most DNS records are A records? May as well hit the most popular rec. type at the beginning:
<img width="1065" alt="image" src="https://user-images.githubusercontent.com/31913027/163780749-a401291a-c983-4865-83f4-7a40d79f627d.png">

I bring this up because I ran into a "on prem" scenario with vendor products. Kong may not be aware, but oddly enough some enterprise DNS/LB solutions out there have their own "bug" because Kong tries to resolve a A record as a SRV first, and somehow that actually causes these tools to at a very low error rate return "A record not found" error code 3's sometimes even when the record exists! Oddly our DNS team claims that sending the SRV prior to the A query is why. I have little confidence in them fixing the vendor product but I just propose Kong to change up the order as it may actually help some enterprise and regular users over time. I made these changes in our own env and now seeing improvement across the board on a few problematic DNS FQDNs internally. 

May even be worth doing: LAST,A,CNAME,SRV , since SRV is likely last on the popularity chain.

**Comment by hbagdi** [comment]
@Tieske knows better but I think the reason for doing this is to _load-balance_ traffic. SRV records have the ability to specify weights and that is something a load-balancer like Kong must take into account.

Isn't the fact that it is configurable not sufficient?

**Reply by jeremyjpj0916** [reply]
Sufficient yes, but I do like apps to be optimized per default configs when it makes sense. Specially in my case what I saw on prem with our 3 private datacenters was the underlying technologies handling dns resolution do not like invalid types being sent prior to a valid type record per a given FQDN on rare occasion and it causes issues (we proved it out with raw packet capture tracing Kong sending SRV first before A rec query caused a error code 3(dns rec not found) about .0004% of the time for 3-4 FQDNs out of 1000s we reverse proxy to). Definitely a bug on the vendor stuff that will likely never get fixed. Being that most of Kongs customers probably don't own their own datacenters/infra they may never hit this case actually.

Can the same FQDN have both a SRV AND an A record stored in a given nameserver? Or can only the FQDN have an A OR SRV record assigned from a given NS. I think the answer to that determines if I personally think putting A record in front of SRV to be a guaranteed worthwhile change since if it can't share two record types, with A being the most popular record type out there so makes logical sense to query by whats most likely to get a hit first. Then follow up with second most popular type(which is probably CNAME alias or so, at least for us it is), etc. 

If a FQDN can share both a A and SRV record from same nameserver then it makes sense as you said for SRV to come first to do fancier things around LBing w DNS results if thats a feature Kong supports.

**Reply by Tieske** [reply]
in automated environments SRV records are used heavily for service discovery (k8s anyone?), and in many of those envs you'll find both SRV and A records for the same name. To make it work reliably the SRV has to go first for weight and port information.

I get your point @jeremyjpj0916 , but it would break a lot of environments that currently work out of the box.

**Reply by jeremyjpj0916** [reply]
Ah if they can share a name then yep makes perfect sense for SRV to come first to support that. More harm that good could come from switching the default then potentially. Case closed 😄. We gave the other dns vendor products the udp packet proof to let them figure out why SRV query prior to A rec query breaks their stuff on super rare occasion so they can fix it themselves. Thanks!
